agent/src/main/java/com/appland/appmap/util/GitUtil.java

```diff
@@ -0,0 +1,237 @@
+      return ret;
+    } catch (NoWorkTreeException | GitAPIException e) {
+      logger.warn(e);
+    }
+    return ret;
+  }
+
+  // In the code below, the "fs" prefix indicates that the variable holds a
+  // filesystem path. The "git" prefix indicates that the variable holds a git
+  // path. Hopefully this will make mistakes easier to spot.
+
+  private static final byte[] GIT_MAIN_JAVA = "src/main/java".getBytes();
+  private static int GIT_MAIN_JAVA_LEN = GIT_MAIN_JAVA.length;
+  private static final byte[] GIT_TEST_JAVA = "src/test/java".getBytes();
+  private static int GIT_TEST_JAVA_LEN = GIT_TEST_JAVA.length;
+
+  /**
+   * If we're in a git repo, walk all the directories under the directory that
+   * contains the config, looking for source directories (i.e. those that staart
+   * "src/main/java" or "src/test/java").
+   *
+   * If we're not in a git repo, do nothing.
+   *
+   * @throws IOException if something goes wrong accessing the git repo
+   */
+  public static void findSourceRoots() throws IOException {
+    // Start the search for source roots in the directory containing the config.
+    Path fsBase = AppMapConfig.get().configFile.toAbsolutePath().getParent();
+
+    try (GitUtil git = GitUtil.open()) {
+      if (git == null) {
+        return; // not in a repo, nothing to do
+      }
+
+      Repository repository = git.getRepository();
+      Path fsRoot = repository.getWorkTree().toPath();
+      String gitBase = fsRoot.relativize(fsBase).toString().replace(File.separator, "/");
+      logger.debug("repoRoot: {}, gitCwd: {}", fsRoot, gitBase);
+      try (TreeWalk treeWalk = new TreeWalk(repository)) {
+        treeWalk.addTree(repository.resolve(Constants.HEAD + "^{tree}"));
+
+        if (gitBase.length() > 0) {
+          // Ignore the parts of the repo that aren't under the base directory.
+          treeWalk.setFilter(PathFilter.create(gitBase));
+        }
+
+        // treeWalk.next() returns directories iff setRecursive(false)
+        treeWalk.setRecursive(false);
+
+        while (treeWalk.next()) {
+          if (treeWalk.isPathSuffix(GIT_MAIN_JAVA, GIT_MAIN_JAVA_LEN)
+              || treeWalk.isPathSuffix(GIT_TEST_JAVA, GIT_TEST_JAVA_LEN)) {
+            String gitSourceRoot = treeWalk.getPathString();
+
+            // Take advantage of the fact that Path treats '/' as a valid
+            // separator, even on Windows:
+            Path fsSourceRoot = fsRoot.resolve(Paths.get(gitSourceRoot));
+
+            // Once we've determined the full git path, strip off the base (if
+            // necessary):
+            if (gitBase.length() > 0) {
+              gitSourceRoot = relativizeGitPath(gitBase, gitSourceRoot);
+            }
+
+            logger.debug("found {} => {}", gitSourceRoot, fsSourceRoot);
+            sourceRoots.put(gitSourceRoot, fsSourceRoot);
+          }
+
+          if (treeWalk.isSubtree()) {
```

Comment:
Should this be an `else if` to prune the unnecessary subtrees?

Reply:
Yes, I believe so. Good catch.